docs/source/advanced-usage.rst

```diff
@@ -0,0 +1,68 @@
+Advanced Usage
+==============
+
+This document explains some of the more advanced usage concepts with
+`wsproto`. This is assume you are familiar with `wsproto` and I/O in
+Python.
+
+Back-pressure
+-------------
+
+Back-pressure is an important concept to understand when implementing a
+client/server protocol. This section briefly explains the issue and then
+explains how to handle back-pressure when using `wsproto`.
+
+Imagine that you have a WebSocket server that reads messages from the client,
+does some processing, and then sends a response. What happens if the client
+sends messages faster than the the server can process them? If the incoming
+messages are buffered in memory, then the server will slowly use more and more
+memory, until the OS eventually kills it. This scenario is directly applicable
+to `wsproto`, because every time you call ``receive_data()``, it appends that
+data to an internal buffer.
+
+The slow endpoint needs a way to signal the fast endpoint to stop sending
+messages until the slow endpoint can catch up. This signaling is called
+"back-pressure". As a Sans-IO library, `wsproto` is not responsible for
+network concerns like back-pressure, so that responsibility belongs to your
+network glue code.
+
+Fortunately, TCP has the ability to signal backpressure, and the operating
+system will do that for you automatically—if you follow a few rules! The OS
+buffers all incoming and outgoing network data. Standard Python socket methods
+like ``send()`` and ``recv()`` copy data to and from those OS buffers. For
```

Comment:
... methods, such as `send(...)` and `recv()`, copy data ...

Reply:
Done.